src/Generator/PromotedParameterGenerator.php

```diff
@@ -87,6 +87,14 @@ public static function fromParameterGeneratorWithVisibility(ParameterGenerator $
             );
         }
 
+        if (
+            null !== $type
+            && ($typeObject = $generator->getTypeObject())
+            && $typeObject->getNullable()
+        ) {
+            $type = '?' . $type;
+        }
```

Comment:
Wondering if the `$type` of `ParameterGenerator` could be exposed internally only: we could then get rid of the string type overall

Reply:
@Grundik I don't have a clear mind on the approach here yet, but I'm trying to think of a design that removes `public`, and ideally avoids `protected` (although the inheritance design is already out of the bag)

Reply:
Right, I'm much in "damage control" mode here, since the generator, builder and reflection-alike structure has been merged together in the past, making changes to internal structures lead to major versions :) 

I'll sleep over it, but I'd rather not expose more than is strictly needed for `->generate()` to work.

Reply:
Nono, I'm absolutely not suggesting that we go down the parsing way :)